README.md
solr-proxy
==========

Reverse proxy to make a Solr instance read-only, rejecting requests that have the potential to modify the Solr index.

This is a rewrite of [solr-security-proxy](https://github.com/dergachev/solr-security-proxy) with some bug fixes and additional features.

[![Build Status](https://secure.travis-ci.org/Trott/solr-proxy.png)](http://travis-ci.org/Trott/solr-proxy)

Installation
------------

For use from the command line:

```bash
npm install -g solr-proxy
```

For use in another application:

```bash
npm install solr-proxy
```

Usage
-----

From the command-line:

```bash
solr-proxy
```

Options are:

```
Options:
  --host          Listen on this host         [default: "localhost"]
  --port          Listen on this port         [default: 8008]
  --upstream      Solr backend                [default: "http://localhost:8983"]
  --validPaths    Allowed paths (comma        [default: "/solr/select"]
                  delimited)
  --invalidParams Blocked parameters (comma   [default: "qt,stream"]
                  delimited)
  --validMethods  Allowed HTTP methods (comma [default: "GET"]
                  delimited)
  --maxRows       Maximum rows permitted in a [default: 200]
                  request
  --maxStart      Maximum start offset        [default: 1000]
                  permitted in a request
  --quiet, -q     Do not write messages to STDOUT
  --version, -v   Show version
  --help, -h      Show this message
```

To start the server from your application:

```js
import SolrProxy from 'solr-proxy'
await SolrProxy.start()
```

You can pass a port number as the first argument to `start()`. You may pass a
falsy value (such as `null` or `undefined`) if you wish to use the port number
specified specified in the `listenPort` property in the options object (second
argument). If the port is not specified in either argument, the default value of
`8008` is used.

You can pass an options object as the second argument to `start()`.

```js
const defaultOptions = {
  listenPort: 8008,                 // port to use if `port` is unspecified or invalid
  validHttpMethods: ['GET'],        // all other HTTP methods will be disallowed
  validPaths: ['/solr/select'],     // all other paths will be denied
  invalidParams: ['qt', 'stream'],  // blocks requests with params qt or stream.* (all other params are allowed)
  upstream: 'http://localhost:8983' // proxy to solr at this location
  maxRows: 200,                     // maximum number of rows allowed in a request
  maxStart: 1000                    // maximum start offset allowed in a request
}
```

To enable TLS for your proxy, include an `ssl` object within the `options`
object.

```js
const options = {
  ssl: {
    key: fs.readFileSync('key.pem'),
    cert: fs.readFileSync('cert.pem'),
  }
}
const proxy = await SolrProxy.start(null, options);
```

Default Rules
-------------

solr-proxy has the following default rules:

* Reject any request methods other than GET
* Only allow the `/solr/select` path
* Block requests with `qt` and `stream.*` query parameters.
* Reject requests with `rows` set to more than 200.
* Reject requests with `start` set to more than 1000.


License
-------

MIT

Related Reading
---------------

* [Why use a Solr proxy?](https://github.com/dergachev/solr-security-proxy#user-content-how-it-works)
* [Solr Security Resources](https://github.com/dergachev/solr-security-proxy#user-content-solr-security-resources)


CHANGELOG.md
# [10.0.0](https://github.com/Trott/solr-proxy/compare/v9.1.0...v10.0.0) (2023-09-01)


### chore

* update package-lock.json ([fa02764](https://github.com/Trott/solr-proxy/commit/fa02764fb30a775d5792dc18ce4a33b62e36b2aa))


### BREAKING CHANGES

* Drop support for versions of Node.js older than 16.x

# [9.1.0](https://github.com/Trott/solr-proxy/compare/v9.0.1...v9.1.0) (2022-08-18)


### Features

* add `--host` option to help with containers ([8305612](https://github.com/Trott/solr-proxy/commit/830561235e246a22ea815aa5d436cf8a67fa2a7e))

## [9.0.1](https://github.com/Trott/solr-proxy/compare/v9.0.0...v9.0.1) (2022-05-03)


### Bug Fixes

* push update to remove deprecation warning ([d0dfcd1](https://github.com/Trott/solr-proxy/commit/d0dfcd19f2144d54f9c3f059db2623acdae5c1c8))

# [9.0.0](https://github.com/Trott/solr-proxy/compare/v8.0.0...v9.0.0) (2022-04-18)


### Features

* add TypeScript types ([18b3c15](https://github.com/Trott/solr-proxy/commit/18b3c15ff59d035aa715d1edea46c1d7c9d0d4cf))


### BREAKING CHANGES

* This module is now ESM only.
* Drop support for Node.js 12.x

# [9.0.0](https://github.com/Trott/solr-proxy/compare/v8.0.0...v9.0.0) (2022-04-18)


### Features

* add TypeScript types ([18b3c15](https://github.com/Trott/solr-proxy/commit/18b3c15ff59d035aa715d1edea46c1d7c9d0d4cf))


### BREAKING CHANGES

* This module is now ESM only.
* Drop support for Node.js 12.x

# [8.0.0](https://github.com/Trott/solr-proxy/compare/v7.0.0...v8.0.0) (2022-02-13)


### chore

* use fastify proxy ([14af414](https://github.com/Trott/solr-proxy/commit/14af414dd94218e5634ecf8678b4569153adfeeb))


### BREAKING CHANGES

* Some options have changed.

# [7.0.0](https://github.com/Trott/solr-proxy/compare/v6.0.1...v7.0.0) (2022-02-11)


### chore

* remove debug dependency ([3774dc6](https://github.com/Trott/solr-proxy/commit/3774dc6af0402aea4a9548060cd76935a61ca50e))


### BREAKING CHANGES

* Use NODE_DEBUG=solr-proxy instead of DEBUG=solr-proxy
for verbose output.

## [6.0.1](https://github.com/Trott/solr-proxy/compare/v6.0.0...v6.0.1) (2022-01-14)


### Bug Fixes

* update dependencies ([2ea26a9](https://github.com/Trott/solr-proxy/commit/2ea26a9fd349039da39cfaec2ba546c5791aedc7))

# [6.0.0](https://github.com/Trott/solr-proxy/compare/v5.0.0...v6.0.0) (2021-05-09)


### chore

* drop support for 10.x ([7ba92b6](https://github.com/Trott/solr-proxy/commit/7ba92b67b526a2371a1f5ed703700eb8a594cf62))


### BREAKING CHANGES

* drop support for Node.js 10.x

# [5.0.0](https://github.com/Trott/solr-proxy/compare/v4.1.1...v5.0.0) (2021-01-04)


### config

* use canonical Solr default port for backend ([d581860](https://github.com/Trott/solr-proxy/commit/d5818609380394bd6eb944e340c052f15ab0ff43))


### BREAKING CHANGES

* If the "backend" (i.e., Solr server) port is not
specified, use the default Solr port of 8983 rather than our own default
of 8080.

Fixes: https://github.com/Trott/solr-proxy/issues/11

4.1.1
=====

* Fix alignment for default options in `--help` output.

4.1.0
=====

* Add new option `maxStart` that defaults to `1000`. (Thanks @tuolumne!)
* Drop testing for EOL Node.js 13.x.

4.0.0
=====

* Add new option `maxRows` that defaults to `200`.

3.0.0
=====

* Drop support for EOL Node.js 8.x.

2.1.2
=====

* Update dependencies includeing security fixes. Please update!

2.1.1
=====

* Fix CLI. Shipped without a required library. Whoops. Testing didn't catch it
  because the required library was in `node_modules` due to a testing
  dependency.  Ooooooffffff.

2.1.0
=====

* Add SSL capability to the proxy.

2.0.0
=====

* Node.js 8.x and 10.x support. Drop support for earlier versions.

1.2.0
=====

* Add quiet mode
* Add debug logging capabilities

1.1.2
=====

* improved line wrapping for usage message

1.1.1
=====

* fix misformatted usage message

1.1.0
=====

* --version flag

1.0.1
=====

* 100% test coverage on `index.js` and `lib/`
* documentation updates

1.0.0
=====

* Initial release


lib/cli/argv.ts
import type SolrProxy from '../../index.js'

import fs from 'fs'
import minimist from 'minimist'

interface ProxyOptions {
  listenHost?: string
  upstream?: string
  validHttpMethods?: string[]
  invalidParams?: string[]
  validPaths?: string[]
  maxRows?: number
  maxStart?: number
}

const createProxyOptions = function (argv: minimist.ParsedArgs): ProxyOptions {
  const proxyOptions: ProxyOptions = {}

  if (argv.host != null) {
    proxyOptions.listenHost = argv.host
  }

  if (argv.upstream != null) {
    proxyOptions.upstream = argv.upstream
  }

  if (argv.validMethods != null) {
    proxyOptions.validHttpMethods = argv.validMethods.split(',')
  }

  if (argv.invalidParams != null) {
    proxyOptions.invalidParams = argv.invalidParams.split(',')
  }

  if (argv.validPaths != null) {
    proxyOptions.validPaths = argv.validPaths.split(',')
  }

  if (argv.maxRows != null) {
    proxyOptions.maxRows = +argv.maxRows
  }

  if (argv.maxStart != null) {
    proxyOptions.maxStart = +argv.maxStart
  }

  return proxyOptions
}

export default async function (argv: minimist.ParsedArgs, stdout: Function, proxy: typeof SolrProxy): Promise<void> {
  const usageMessage = 'Usage: solr-proxy [options]\n' +
    '\n' +
   'Options:\n' +
   '--host          Listen on this host         [default: "localhost"]\n' +
   '--port          Listen on this port         [default: 8008]\n' +
   '--upstream      Solr backend                [default: "http://localhost:8983"]\n' +
   '--validPaths    Allowed paths (comma        [default: "/solr/select"]\n' +
   '                delimited)\n' +
   '--invalidParams Blocked parameters (comma   [default: "qt,stream"]\n' +
   '                delimited)\n' +
   '--validMethods  Allowed HTTP methods (comma [default: "GET"]\n' +
   '                delimited)\n' +
   '--maxRows       Maximum rows permitted in a [default: 200]\n' +
   '                request\n' +
   '--maxStart      Maximum start offset        [default: 1000]\n' +
   '                permitted in a request\n' +
   '--quiet, -q     Do not write messages to STDOUT\n' +
   '--version, -v   Show version\n' +
   '--help, -h      Show this message'

  if (argv.help != null || argv.h != null) {
    stdout(usageMessage)
    return
  }

  if (argv.version != null || argv.v != null) {
    const version = JSON.parse(fs.readFileSync(new URL('../../package.json', import.meta.url), 'utf-8')).version
    stdout(version)
    return
  }

  if (argv.quiet != null || argv.q != null) {
    // quiet mode, change stdout to a no-op
    stdout = function () {}
  }

  const proxyOptions = createProxyOptions(argv)

  await proxy.start(argv.port, proxyOptions)
  stdout('solr-proxy is running...')
}
